subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c

```diff
@@ -4,103 +4,133 @@
+	static struct ec_host_cmd_espi_ctx _name##_hc_espi;                                        \
+	struct ec_host_cmd_backend _name = {                                                       \
+		.api = &ec_host_cmd_api,                                                           \
+		.ctx = (struct ec_host_cmd_espi_ctx *)&_name##_hc_espi,                            \
+	}
+
+static void espi_handler(const struct device *dev, struct espi_callback *cb,
+			 struct espi_event espi_evt)
 {
-	struct ec_host_cmd_backend_espi_data *data =
-		CONTAINER_OF(cb, struct ec_host_cmd_backend_espi_data, espi_cb);
+	struct ec_host_cmd_espi_ctx *hc_espi =
+		CONTAINER_OF(cb, struct ec_host_cmd_espi_ctx, espi_cb);
 	uint16_t event_type = (uint16_t)espi_evt.evt_details;
+	/* tx stores the shared memory buf pointer and size, so use it */
+	const struct ec_host_cmd_request_header *rx_header = hc_espi->tx->buf;
+	const size_t shared_size = hc_espi->tx->len_max;
+	const uint16_t rx_valid_data_size = rx_header->data_len + RX_HEADER_SIZE;
 
 	if (event_type != ESPI_PERIPHERAL_EC_HOST_CMD) {
 		return;
 	}
 
-	if (k_sem_take(&data->dev_owns, K_NO_WAIT) != 0) {
-		int res = EC_HOST_CMD_IN_PROGRESS;
-
-		espi_write_lpc_request(ESPI_DEVICE, ECUSTOM_HOST_CMD_SEND_RESULT, &res);
-		return;
+	/* Only support version 3 and make sure the number of bytes to copy is not
+	 * bigger than rx buf size or the shared memory size
+	 */
+	if (rx_header->prtcl_ver != 3 ||
+	    rx_valid_data_size > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER ||
+	    rx_valid_data_size > shared_size) {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, RX_HEADER_SIZE);
+		hc_espi->rx_ctx->len = RX_HEADER_SIZE;
+	} else {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, rx_valid_data_size);
+		hc_espi->rx_ctx->len = rx_valid_data_size;
 	}
 
-	k_sem_give(&data->handler_owns);
+	/* Even in case of errors, let the general handler send response */
+	k_sem_give(&hc_espi->rx_ctx->handler_owns);
 }
 
-int ec_host_cmd_backend_espi_init(const struct device *dev, struct ec_host_cmd_rx_ctx *rx_ctx)
+static int ec_host_cmd_espi_init(const struct ec_host_cmd_backend *backend, const void *config,
+				 struct ec_host_cmd_rx_ctx *rx_ctx, struct ec_host_cmd_tx_buf *tx)
 {
-	struct ec_host_cmd_backend_espi_data *data = dev->data;
+	struct ec_host_cmd_espi_ctx *hc_espi = (struct ec_host_cmd_espi_ctx *)backend->ctx;
 
-	if (rx_ctx == NULL) {
-		return -EINVAL;
+	hc_espi->espi_dev = ((const struct ec_host_cmd_config *)config)->dev;
+
+	if (!device_is_ready(hc_espi->espi_dev)) {
+		return -ENODEV;
+	}
+
+	/* eSPI backend needs rx buffer provided by the handler */
+	if (!rx_ctx->buf) {
+		return -EIO;
 	}
 
-	rx_ctx->buf = data->espi_shm;
-	rx_ctx->len = &data->rx_buffer_len;
-	rx_ctx->dev_owns = &data->dev_owns;
-	rx_ctx->handler_owns = &data->handler_owns;
+	/* eSPI backend uses the shared memory as tx buffer */
+	if (tx->buf) {
+		LOG_WRN("Providing unused buffer");
```

Comment:
Move this check (and other backends) into ec_host_cmd_init() so you have a consistent warning message regardless of the backend selected.